zetteldeft.el

```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
   :type 'string
   :group 'zetteldeft)
 
-(defun zetteldeft--insert-title ()
-  "Insert filename of current zd note, stripped from its ID.
+(defun zetteldeft--insert-title (title)
+  "Insert TITLE as title in file.
 Prepended by `zetteldeft-title-prefix' and appended by `zetteldeft-title-suffix'."
   (zetteldeft--check)
   (insert
     zetteldeft-title-prefix
-    (zetteldeft--lift-file-title (file-name-base (buffer-file-name)))
+    title
     zetteldeft-title-suffix))
 
 (defun zetteldeft--lift-file-title (zdFile)
   "Return the title of a zetteldeft note.
 ZDFILE should be a full path to a note."
-  (let ((baseName (file-name-base zdFile)))
-    (replace-regexp-in-string
-     (concat zetteldeft-id-regex "[[:space:]]")
-     "" baseName)))
-
-(defun zetteldeft-file-rename ()
-  "Rename the current file via the deft function.
-Use this on files in the `deft-directory'."
-  (interactive)
+  (let ((deft-use-filename-as-title nil)
+	contents title)
+    (with-current-buffer (get-buffer-create "*Zetteldeft temp*")
+      (insert-file-contents zdFile nil nil nil t)
+      (setq contents (concat (buffer-string))))
+    (setq title (deft-parse-title zdFile contents))
+    (kill-buffer "*Zetteldeft temp*")
+    title))
+
+(defun zetteldeft--rename (cusprompt fn-gen update-title)
+  "Generic function to rename a file.
+Takes CUSPROMPT, FN-GEN and UPDATE-TITLE to customise behaviour."
   (zetteldeft--check)
-    (let ((old-filename (buffer-file-name))
-          (deft-dir (file-name-as-directory deft-directory))
-          new-filename old-name new-name)
-      (when old-filename
-        (setq old-name (deft-base-filename old-filename))
-        (setq new-name (read-string
-                        (concat "Rename " old-name " to (without extension): ")
-                        old-name))
-        (setq new-filename
-              (concat deft-dir new-name "." deft-default-extension))
-        (rename-file old-filename new-filename)
-        (deft-update-visiting-buffers old-filename new-filename)
-        (zetteldeft-update-title-in-file)
-        (deft-refresh))))
-
-(defun zetteldeft-update-title-in-file ()
+  (let ((old-filename (buffer-file-name))
+	prompt-pars new-name new-filename)
+    (when old-filename
+      (setq prompt-pars (funcall cusprompt old-filename))
+      (setq new-name (read-string (car prompt-pars) (cdr prompt-pars)))
+      (setq new-filename (funcall fn-gen old-filename new-name))
```

Comment:
Creating unassigned variables and then calling `setq` is awkward. Prefer `let*`.

Reply:
Thanks - updated!